From b2b10b426b292e8dbb0789b19f839f3cecdce557 Mon Sep 17 00:00:00 2001
From: NLObP <38990417+NL0bP@users.noreply.github.com>
Date: Tue, 6 Aug 2024 22:32:45 +0300
Subject: [PATCH] [Game] Fixed starting quests launching: - now there are no
 erroneous entries in the quest list;

[Game] Added send SCScheduleItemUpdatePacket:
- Divine clock is still not working.
---
 AAEmu.Game/Core/Managers/TimedRewardsManager.cs  |  2 +-
 AAEmu.Game/Core/Managers/World/WorldManager.cs   |  9 ++++++++-
 .../Core/Packets/C2G/CSSelectCharacterPacket.cs  | 16 ++++------------
 .../Core/Packets/Proxy/FinishStatePacket.cs      |  8 ++++----
 .../Models/Game/Items/Actions/ItemTaskType.cs    |  3 ++-
 AAEmu.Game/Models/Game/ScheduleItem.cs           |  2 +-
 .../Characters/CharacterOnlineTrackingTask.cs    | 14 ++++++++++++++
 7 files changed, 34 insertions(+), 20 deletions(-)

diff --git a/AAEmu.Game/Core/Managers/TimedRewardsManager.cs b/AAEmu.Game/Core/Managers/TimedRewardsManager.cs
index db416388eb..0eaa9ed643 100644
--- a/AAEmu.Game/Core/Managers/TimedRewardsManager.cs
+++ b/AAEmu.Game/Core/Managers/TimedRewardsManager.cs
@@ -97,7 +97,7 @@ public void DoDailyAccountLogin(ulong accountId)
 
         if (AppConfiguration.Instance.Loyalty.DailyLogin > 0)
             AccountManager.Instance.AddLoyalty(accountId, AppConfiguration.Instance.Loyalty.DailyLogin);
-        
+
         AccountManager.Instance.UpdateDivineClock(accountId, 0, 0);
     }
 
diff --git a/AAEmu.Game/Core/Managers/World/WorldManager.cs b/AAEmu.Game/Core/Managers/World/WorldManager.cs
index 2c51b4c65e..a5a2aeb670 100644
--- a/AAEmu.Game/Core/Managers/World/WorldManager.cs
+++ b/AAEmu.Game/Core/Managers/World/WorldManager.cs
@@ -1240,25 +1240,32 @@ public void OnPlayerJoin(Character character)
     private void StartingFirstJourney(Character character)
     {
         var questId = 0u;
+        var sphereId = 0u;
         switch (character.Race)
         {
             case Race.Nuian: // Nuian
                 questId = 6839;
+                sphereId = 2321;
                 break;
             case Race.Dwarf: // Dwarf
                 questId = 5811;
+                sphereId = 2613;
                 break;
             case Race.Elf: // Elf
                 questId = 6840;
+                sphereId = 2322;
                 break;
             case Race.Hariharan: // Hariharan
                 questId = 6842;
+                sphereId = 2324;
                 break;
             case Race.Ferre: // Ferre
                 questId = 6841;
+                sphereId = 2323;
                 break;
             case Race.Warborn: // Warborn
                 questId = 8228;
+                sphereId = 2600;
                 break;
             case Race.Fairy:
                 break;
@@ -1268,7 +1275,7 @@ private void StartingFirstJourney(Character character)
         // showing it once
         if (character.Updated - character.Created < TimeSpan.FromMinutes(1))
         {
-            character.Quests.AddQuest(questId);
+            character.Quests.AddQuestFromSphere(questId, sphereId);
         }
     }
 
diff --git a/AAEmu.Game/Core/Packets/C2G/CSSelectCharacterPacket.cs b/AAEmu.Game/Core/Packets/C2G/CSSelectCharacterPacket.cs
index 09749330e4..afc5d24638 100644
--- a/AAEmu.Game/Core/Packets/C2G/CSSelectCharacterPacket.cs
+++ b/AAEmu.Game/Core/Packets/C2G/CSSelectCharacterPacket.cs
@@ -9,6 +9,7 @@
 using AAEmu.Game.Core.Network.Game;
 using AAEmu.Game.Core.Packets.G2C;
 using AAEmu.Game.Models.Game.Char;
+using AAEmu.Game.Models.Game.Items.Actions;
 using AAEmu.Game.Models.Game.Skills;
 using AAEmu.Game.Models.Game.Units;
 using AAEmu.Game.Models.Game.Units.Route;
@@ -64,18 +65,8 @@ public override void Read(PacketStream stream)
             Connection.ActiveChar.Simulation = new Simulation(character);
 
             // TODO: Fix the patron and auction house license buff issue
-            if (Connection.ActiveChar is { } unit)
-            {
-                if (!unit.Buffs.CheckBuff((uint)SkillConstants.Patron)) //TODO Wrong place
-                {
-                    unit.Buffs.AddBuff(new Buff(unit, unit, SkillCaster.GetByType(SkillCasterType.Unit), SkillManager.Instance.GetBuffTemplate(8000011), null, System.DateTime.Now));
-                }
-
-                if (!unit.Buffs.CheckBuff((uint)SkillConstants.AuctionLicense)) //TODO Wrong place
-                {
-                    unit.Buffs.AddBuff(new Buff(unit, unit, SkillCaster.GetByType(SkillCasterType.Unit), SkillManager.Instance.GetBuffTemplate(8000012), null, System.DateTime.Now));
-                }
-            }
+            Connection.ActiveChar.Buffs.AddBuff((uint)SkillConstants.Patron, Connection.ActiveChar);
+            Connection.ActiveChar.Buffs.AddBuff((uint)SkillConstants.AuctionLicense, Connection.ActiveChar);
 
             Connection.SendPacket(new SCResidentInfoListPacket(ResidentManager.Instance.GetInfo()));
             Connection.SendPacket(new SCCharacterStatePacket(character));
@@ -126,6 +117,7 @@ public override void Read(PacketStream stream)
             character.Breath = character.LungCapacity;
 
             Connection.SendPacket(new SCScheduledEventStartedPacket());
+            Connection.SendPacket(new SCItemTaskSuccessPacket(ItemTaskType.Unk136, [], []));
 
             Connection.ActiveChar.OnZoneChange(0, Connection.ActiveChar.Transform.ZoneId);
         }
diff --git a/AAEmu.Game/Core/Packets/Proxy/FinishStatePacket.cs b/AAEmu.Game/Core/Packets/Proxy/FinishStatePacket.cs
index ff0a5f14eb..135df82c33 100644
--- a/AAEmu.Game/Core/Packets/Proxy/FinishStatePacket.cs
+++ b/AAEmu.Game/Core/Packets/Proxy/FinishStatePacket.cs
@@ -38,9 +38,9 @@ public override void Read(PacketStream stream)
                 // Test URLs                                          // Original Trion values
                 // Client treats these as folders and will add a trailing slash (/) with whatever it needs
                 // For example, opening the Wiki would send http://localhost/aaemu/platform/login
-                var authUrl = "http://localhost/aaemu/login";         // "https://session.draft.integration.triongames.priv";
-                var platformUrl = "http://localhost/aaemu/platform";  // "http://archeage.draft.integration.triongames.priv/commerce/pruchase/credits/purchase-credits-flow.action";
-                var commerceUrl = "http://localhost/aaemu/shop";      // "" ;
+                //var authUrl = "http://localhost/aaemu/login";         // "https://session.draft.integration.triongames.priv";
+                //var platformUrl = "http://localhost/aaemu/platform";  // "http://archeage.draft.integration.triongames.priv/commerce/pruchase/credits/purchase-credits-flow.action";
+                //var commerceUrl = "http://localhost/aaemu/shop";      // "" ;
 
                 // It seems this packet can be ignored if you don't use the wiki/shop
                 //Connection.SendPacket(new SCTrionConfigPacket(
@@ -56,7 +56,7 @@ public override void Read(PacketStream stream)
                         Connection.Payment.EndTime)
                 );
                 Connection.SendPacket(new SCChatSpamConfigPacket());
-                Connection.SendPacket(new SCAccountAttributeConfigPacket([false, false, false]));
+                Connection.SendPacket(new SCAccountAttributeConfigPacket([false, true, false]));
                 Connection.SendPacket(new SCLevelRestrictionConfigPacket(0, 10, 0, 10, 10, [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])); // TODO - config files
 
                 Connection.SendPacket(new SCTaxItemConfigPacket(0));
diff --git a/AAEmu.Game/Models/Game/Items/Actions/ItemTaskType.cs b/AAEmu.Game/Models/Game/Items/Actions/ItemTaskType.cs
index 141a16e831..3405611af2 100644
--- a/AAEmu.Game/Models/Game/Items/Actions/ItemTaskType.cs
+++ b/AAEmu.Game/Models/Game/Items/Actions/ItemTaskType.cs
@@ -119,5 +119,6 @@ public enum ItemTaskType : byte
     Sknize = 113,
     ItemTaskThistimeUnpack = 114,
     BuyPremiumService = 115,
-    BuyAaPoint = 116
+    BuyAaPoint = 116,
+    Unk136 = 136
 }
diff --git a/AAEmu.Game/Models/Game/ScheduleItem.cs b/AAEmu.Game/Models/Game/ScheduleItem.cs
index 39a5a81f89..b4d68f3197 100644
--- a/AAEmu.Game/Models/Game/ScheduleItem.cs
+++ b/AAEmu.Game/Models/Game/ScheduleItem.cs
@@ -7,7 +7,7 @@ public class ScheduleItem : PacketMarshaler
 {
     public uint ItemTemplateId { get; set; }
     public byte Gave { get; set; }
-    public int Acumulated { get; set; }
+    public uint Acumulated { get; set; }
     public DateTime Updated { get; set; }
 
     public override PacketStream Write(PacketStream stream)
diff --git a/AAEmu.Game/Models/Tasks/Characters/CharacterOnlineTrackingTask.cs b/AAEmu.Game/Models/Tasks/Characters/CharacterOnlineTrackingTask.cs
index b59e5e7850..e7fbe58667 100644
--- a/AAEmu.Game/Models/Tasks/Characters/CharacterOnlineTrackingTask.cs
+++ b/AAEmu.Game/Models/Tasks/Characters/CharacterOnlineTrackingTask.cs
@@ -1,6 +1,8 @@
 ﻿using System;
 using AAEmu.Game.Core.Managers;
 using AAEmu.Game.Core.Managers.World;
+using AAEmu.Game.Core.Packets.G2C;
+using AAEmu.Game.Models.Game;
 
 namespace AAEmu.Game.Models.Tasks.Characters;
 
@@ -45,6 +47,18 @@ public override void Execute()
 
             // TODO: Use lastSeconds and newSeconds as a comparison for triggering time played achievements
             // TODO: Add divine clock feedback packets
+
+            if (time % 60 == 0)
+            {
+                var si = new ScheduleItem
+                {
+                    ItemTemplateId = 9000003,
+                    Gave = (byte)taken,
+                    Acumulated = time,
+                    Updated = DateTime.UtcNow
+                };
+                character.SendPacket(new SCScheduleItemUpdatePacket([si]));
+            }
         }
 
         lock (_lock)
